### ISSUE TITLE
Nested catalog merge optimisation

### ISSUE THREAD

**Issue by mcb30** [cvmfs/cvmfs#2707]
I have done some investigation of using `cvmfs-gateway` to handle large volumes of grafted files.  My experiments suggest that we are currently missing a key optimisation that could increase the throughput by a factor of a hundred or more.

In `catalog_diff_tool_impl.h` there exists an optimisation:
https://github.com/cvmfs/cvmfs/blob/8e0712b0230afe859fc7b5bd421878b4d64677f8/cvmfs/catalog_diff_tool_impl.h#L173-L178

The above optimisation eliminates the need to recurse into nested catalog mountpoints to find differences when the nested catalogs to be compared are known to be identical.  With a well-structured repository, this can eliminate 99% of the work required to find the difference between old and new catalogs.

A similar optimisation is available when performing a merge operation.  At the point that the `ReportModification()` callback is called: if both `entry1` and `entry2` are nested catalog mountpoints, then we could just call `UpdateNestedCatalog()` to record the new hash value for the nested catalog.  Having done this, there would be no need to recurse into the nested catalog mountpoints, because the act of updating the hash would have already performed all of the required changes.  Something along the lines of patching:
https://github.com/cvmfs/cvmfs/blob/8e0712b0230afe859fc7b5bd421878b4d64677f8/cvmfs/receiver/catalog_merge_tool_impl.h#L189-L201
to become:
```c++
  if (entry1.IsDirectory() && entry2.IsDirectory()) {
    // From directory to directory
    const catalog::DirectoryEntryBase* base_entry =
        static_cast<const catalog::DirectoryEntryBase*>(&entry2);
    output_catalog_mgr_->TouchDirectory(*base_entry, xattrs, rel_path.c_str());
    if (entry1.IsNestedCatalogMountpoint() &&
        entry2.IsNestedCatalogMountpoint()) {
      // ... get new hash and size from entry2 ...
      output_catalog_mgr_->UpdateNestedCatalog(std::string(rel_path.c_str()),
                                               new_hash, new_size);
      return true; // skip recursion into nested catalog mountpoints
    } else if (!entry1.IsNestedCatalogMountpoint() &&
        entry2.IsNestedCatalogMountpoint()) {
      output_catalog_mgr_->CreateNestedCatalog(std::string(rel_path.c_str()));
    } else if (entry1.IsNestedCatalogMountpoint() &&
               !entry2.IsNestedCatalogMountpoint()) {
      output_catalog_mgr_->RemoveNestedCatalog(std::string(rel_path.c_str()));
    }
    perf::Inc(counters_->n_directories_changed);
  } ...
```

There is a `WritableCatalog::UpdateNestedCatalog()` that appears to include the logic for updating the `nested_catalogs` table, but it's not clear to me how to obtain the `hash`, `size` and `child_counters` values from `entry2`.

The `ReportModification` callback would also need to be modified to allow a return value to indicate whether or not to continue recursing, but that is a very straightforward change.

To put some concrete performance numbers around this: on my test system, when using `cvmfs_swissknife sync` without a gateway, I measure:
* 22,000 files per second (45 µs/file) for filesystem traversal and graft file parsing
* 50,000 files per second (20 µs/file) for SQLite3 database insertion
* 15,000 files per second (65 µs/file) combined

However, when running the gateway instead (on the same test hardware), I measure only:
* 5,000 files per second (200 µs/file) for catalog diff and merge

The gateway is therefore slowing down, rather than speeding up, the overall publication process, and this seems to be due solely to the overhead of the recursive merge.  With the suggested catalog merge optimisation (and a suitably structured repository), the time taken for the catalog diff and merge should become O(1) rather than O(n) in the number of changed files, and the gateway would no longer be the bottleneck.

Questions:

* What is the correct way to get the `hash`, `size`, and `child_counters` values from `entry2` for the eventual call to `WritableCatalog::UpdateNestedCatalog()`?
* All the existing code in `CatalogMergeTool<RwCatalogMgr, RoCatalogMgr>::ReportModification()` uses `output_catalog_mgr_` to effect changes to the output catalog.  This suggests that a `WriteableCatalogManager::UpdateNestedCatalog()` method should be created, rather than having `ReportModification()` call `WritableCatalog::UpdateNestedCatalog` directly.  Stylistically, what should the parameters look like for this `WriteableCatalogManager::UpdateNestedCatalog()` method?
* Are there any other actions that would need to be taken to ensure a correct outcome (e.g. updating counters, dealing with inode numbers, etc)?
* Could anyone help implement this?  It should massively improve the gateway throughput, but I'm not confident that I understand the catalog internals well enough to implement it myself.

**Comment by mcb30** [comment]
> The `ReportModification` callback would also need to be modified to allow a return value to indicate whether or not to continue recursing, but that is a very straightforward change.

This trivial part is ready, at least: https://github.com/mcb30/cvmfs/tree/nested-merge

**Comment by jblomer** [comment]
Many thanks for bringing this up and for the detailed analysis!  I think that's a great idea and the approach looks viable to me!

I think we should call the new method `WritableCatalogManager::SwapNestedCatalog()` (instead of `UpdateNestedCatalog()`).  This method can call `WritableCatalog::RemoveNestedCatalog` followed by `WritableCatalog::InsertNestedCatalog`. It also needs to fix-up the catalog counters in the parent catalog, so it needs to have a `Catalog` object of the new nested catalog in addition to its hash and size. I think we should require that the nested subtree is _not_ already attached when calling `SwapNestedCatalog()`. Otherwise the logic would become more complicated and we know that the subtree is so-far unseen when we reach it from the receiver's catalog merge tool.

I think we also want to ensure that the new nested catalog won't be attached during `WritableCatalogManager::SwapNestedCatalog()`, otherwise it would be re-compressed and uploaded on commit. We probably get by by only passing the path (nested catalog mountpoint) and the new hash to `SwapNestedCatalog()`. The new hash we get from `LookupNested()` in `new_catalog_mgr_`, for which we should introduce a getter.  The `SwapNestedCatalog()` method can then use the `LoadCatalog()` method from the base class to download and decompress the nested catalog. We can look at the resulting file for the size (perhaps also cross-check with the size reported by `LookupNested()`).  We can then create a new, free-floating `Catalog` object in order to get access to the catalog counters.

Since this is a change deep in the guts of cvmfs, the patch should come with some new tests. I think we should have unit tests for the new `WritableCatalogManager::SwapNestedCatalog()` method in `t_catalog_mgr_rw.cc`. And integration tests where we test swapping different "structure combinations", such as an empty catalog replaced by a complex one with nested children and grand-children and vice versa. We can use `cvmfs_server check` in the integration tests to verify sane repository catalogs after publish operations.

**Comment by jblomer** [comment]
I realize that for the catalog counters fix-up we need both the old nested catalog and the new nested catalog so that we can calculate the correct diff.

**Comment by mcb30** [comment]
@jblomer I've created a draft PR #2709 with a first draft of the code (excluding tests).  Could you let me know if this is along the lines you expected?

I found it quite difficult to understand the separation of responsibilities between various portions of the existing code, so I apologise in advance if this is completely wrong.

**Comment by jblomer** [comment]
Sorry for being a bit unresponsive, this week is [CHEP](https://indico.cern.ch/event/948465), one of our biggest conferences. I'll get back to the PR as soon as possible.

**Comment by mcb30** [comment]
Fixed in #2709
